test/balancerd/mzcompose.py

```diff
@@ -0,0 +1,187 @@
+    rows = cursor.fetchall()
+    assert len(rows) == 1
+    cols = rows[0]
+    assert len(cols) == 1
+    col = cols[0]
+    assert len(col) == (1024 * 1024 * 96) + (2 * 3)
+    assert col.startswith("ABCx")
+    assert col.endswith("xXYZ")
+
+
+def workflow_long_result(c: Composition) -> None:
+    """Test passthrough of long results"""
+    c.up("balancerd", "materialized")
+
+    cursor = c.sql_cursor(service="balancerd")
+    cursor.execute(
+        "SELECT 'ABC', generate_series, 'XYZ' FROM generate_series(1, 10 * 1024 * 1024)"
+    )
+    cnt = 0
+    for row in cursor.fetchall():
+        cnt = cnt + 1
+        assert len(row) == 3
+        assert row[0] == "ABC"
+        assert row[2] == "XYZ"
+    assert cnt == 10 * 1024 * 1024
+
+
+def workflow_long_query(c: Composition) -> None:
+    """Test passthrough of a long SQL query."""
+    c.up("balancerd", "materialized")
+
+    cursor = c.sql_cursor(service="balancerd")
+    small_pad_size = 512 * 1024
+    small_pad = "x" * small_pad_size
+    cursor.execute(f"SELECT 'ABC{small_pad}XYZ';")
+    rows = cursor.fetchall()
+    assert len(rows) == 1
+    cols = rows[0]
+    assert len(cols) == 1
+    col = cols[0]
+    assert len(col) == small_pad_size + (2 * 3)
+    assert col.startswith("ABCx")
+    assert col.endswith("xXYZ")
+
+    medium_pad_size = 1 * 1024 * 1024
+    medium_pad = "x" * medium_pad_size
+    try:
+        cursor.execute(f"SELECT 'ABC{medium_pad}XYZ';")
+        assert False, "execute() expected to fail"
+    except ProgrammingError as e:
+        assert "statement batch size cannot exceed 1000.0 KB" in str(e)
+    except:
+        assert False, "execute() threw an unexpected exception"
+
+    large_pad_size = 512 * 1024 * 1024
+    large_pad = "x" * large_pad_size
+    try:
+        cursor.execute(f"SELECT 'ABC{large_pad}XYZ';")
+        assert False, "execute() expected to fail"
+    except InterfaceError as e:
+        assert "network error" in str(e)
+    except:
+        assert False, "execute() threw an unexpected exception"
+
+    # Confirm that balancerd remains up
+    cursor = c.sql_cursor(service="balancerd")
+    cursor.execute("SELECT 1;")
+
+
+def workflow_mz_restarted(c: Composition) -> None:
+    """Existing connections should fail if Mz is restarted"""
+    c.up("materialized", "balancerd")
+
+    cursor = c.sql_cursor(service="balancerd")
+
+    cursor.execute("CREATE TABLE restart (f1 INTEGER)")
+    cursor.execute("START TRANSACTION")
+    cursor.execute("INSERT INTO restart VALUES (1)")
+    c.kill("materialized")
+    c.up("materialized")
+    try:
+        cursor.execute("INSERT INTO restart VALUES (2)")
+        assert False, "execute() expected to fail"
+    except InterfaceError as e:
+        assert "network error" in str(e)
+    except:
+        assert False, "execute() threw an unexpected exception"
+
+    # Future connections work
+    c.sql_cursor(service="balancerd")
+
+
+def workflow_mz_not_running(c: Composition) -> None:
```

Comment:
Should write this same test but with a down balancerd.

Reply:
Actually this is going to  be too different from production -- if the balancerd is killed, the container dies and then `docker port` fails. So I will not be adding that scenario.